File: Java/FindDuplicates.java
package Java;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class FindDuplicates {
  private static void findDuplicatesInArray(Object[] array) {
    Map<Object, Integer> map = new HashMap<>();
    List<Object> duplicates = new ArrayList<>();

    for (Object obj : array) {
      if (!map.containsKey(obj)) {
        map.put(obj, 1);
      } else {
        map.put(obj, map.get(obj) + 1);
      }
    }

    for (Object key : map.keySet()) {
      if (map.get(key) > 1) {
        duplicates.add(key);
      }
    }
    if (duplicates.isEmpty()) System.out.println("No duplicates!");
    else System.out.println("Duplicates: " + duplicates);
  } 

  public static void main(String[] args) {
    Integer[] nums = {1, 2, 3, 2, 4, 5, 1};
    Integer[] numsUnique = {10, 20, 30, 40, 50};
    Integer[] numsSame = {7,7,7,7,7};
    Integer[] numsEmpty = {};
    String[] words = {"apple", "banana", "apple", "orange", "banana", "kiwi"};

    findDuplicatesInArray(nums);
    findDuplicatesInArray(numsUnique);
    findDuplicatesInArray(numsSame);
    findDuplicatesInArray(numsEmpty);
    findDuplicatesInArray(words);

  }
}

File: TS/findDuplicates.ts
function findDuplicates(array: any[]): any[] | string {
  let map = new Map<any, number>();
  let duplicates = [];

  for (let item of array) {
    if (!map.has(item)) {
      map.set(item, 1);
    } else {
      map.set(item, map.get(item)! + 1);
    }
  }

  for (let key of map.keys()) {
    if (map.get(key)! > 1) {
      duplicates.push(key);
    }
  }
  if (duplicates.length > 0) return duplicates;
  else return "No duplicates"
}

console.log(findDuplicates([1, 2, 3, 2, 4, 5, 1]));
console.log(findDuplicates([10, 20, 30, 40, 50]));
console.log(findDuplicates([7,7,7,7,7]));
console.log(findDuplicates([]));
console.log(findDuplicates(["apple", "banana", "apple", "orange", "banana", "kiwi"]));